sre-recipes/recipes/ratings_freshness_recipe/ratings_freshness_recipe.py

```diff
@@ -0,0 +1,134 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# -*- coding: utf-8 -*-
+"""
+This module contains the implementation of rating_freshness recipe
+"""
+
+import logging
+import subprocess
+from recipe import Recipe
+
+
+class RatingsFreshnessRecipe(Recipe):
+    """
+    This class implements 'rating_freshness' which purposefully
+    stops recollect API calls to rating service.
+    """
+
+    name = "rating_freshness"
+
+    def get_name(self):
+        return self.name
+
+    def is_active(self):
+        return True
+
+    def break_service(self):
+        """
+        Pause Cloud scheduler job to stop calls to recollect API of
+        the rating service
+        """
+        logging.info("Pausing scheduled job of the rating service")
+        print("Breaking rating operations...")
+        project_id = Recipe._get_project_id()
+        if not project_id:
+            print("Failed: cannot find project id.")
+            logging.error("Failed pausing scheduled job: no project id.")
+            exit(1)
+        # hide error output to avoid additional hints :)
+        pause_command = "gcloud scheduler jobs pause ratingservice-recollect-job --project {pid} 2> /dev/null".format(
+            pid=project_id
+        )
+        Recipe._run_command(pause_command)
+        print("...done")
+        logging.info("Scheduled job of the rating service paused")
+
+    def restore_service(self):
+        """
+        Resume Cloud scheduler job to restore calls to recollect API of
+        the rating service
+        """
+        logging.info("Resuming scheduled job of the rating service")
+        print("Restoring rating operations...")
+        project_id = Recipe._get_project_id()
+        if not project_id:
+            print("Failed: cannot find project id.")
+            logging.error("Failed pausing scheduled job: no project id.")
+            exit(1)
+        # hide error output to avoid additional hints :)
+        resume_command = "gcloud scheduler jobs resume ratingservice-recollect-job --project {pid} 2> /dev/null".format(
+            pid=project_id
+        )
+        Recipe._run_command(resume_command)
+        print("...done")
+        logging.info("Scheduled job of the rating service resumed")
+
+    def hint(self):
+        """
+        Provides a hint about the root cause of the issue
+        """
+        project_id = Recipe._get_project_id()
+        print(
+            "Try to change a rating for a shop's items one or more times.",
+            "Try to check SLO Monitoring and look for SLOs that consume error budget.",
+            "For SLO Monitoring open https://pantheon.corp.google.com/monitoring/services?project={pid}".format(
```

Comment:
When I look at the link, I see SLOs for the frontend, recommendationservice, and currencyservice. Is that expected or did something break?

Reply:
I expect that a first line will direct the reader to look into relevant SLOs. Would you like to propose a different wording for the hint?

Reply:
Ok, I just wanted to make sure I was seeing what was expected!

Can you give a quick walkthrough of the thought process you think users will have to debug the issue? I want to make sure there's a clear metric or something they could notice to track down the problem. If the only clue is multiple failing SLOs, it can be hard to find the root cause

Reply:
A starting point can be anything and depends on the background of the investigator, the initial info and setup, etc.
After reading the hint, I expect that a user will try to change rating for one or more products few times and see that it does not refresh neither immediately nor after some time.
After that, a user will go to look which services have problems with SLO or directly go to check error budget graph for rating service SLOs. There, i expect the user to see that error budget for data freshness is in decrease.
There are few possible paths from here but all are based on looking into architecture diagram and understanding what components of the rating service are there. The fact that availability SLO isn't hurt should hint that the problem is not with the service itself and to direct the user to check DB and Scheduler.

Reply:
> After reading the hint, I expect that a user will try to change rating for one or more products few times and see that it does not refresh neither immediately nor after some time.

By refresh, you mean update the score, right? Personally, I'm not sure if this is clear enough. As a user, I'd assume an individual rating just didn't affect the average much

@nicoleczhu do you have thoughts on this?

Reply:
One might not. Also there are few assumptions that, for my opinion, are reasonable to make:
- the long living sandbox environments are rare
- due to configuration user sees changes to the product rating within a minute since last change due to a small initial count of votes (=1) and recurrent call to recollect API (1min)
These assumptions seem reasonable to me. With these assumptions the updating of the product rating even once will have visible result. Also if a rating is 4 and a user selects to update it to 5 then the results will be less visible. But again, I am building on a common sense of the user who knows that the rating is calculating using average (or weighted average) and the new rating selection will be made to change the product's rating faster. At the end, this is a sandbox environment intended for users with a certain background.

Reply:
UX comments while looking at this recipe with new eyes: 

1. Remove the first hint as it detracts from the ultimate goal of learning about SLOs.
2. We should roughly explain the components of ratings service, especially what recollect API does. We shouldn't require users to study app architecture. We're showcasing that you can get "observability" in a single/accessible view, so we shouldn't make users dig into code/READMEs. 
3. Given that we made issue identification easer for the user, your quiz questions could actually be a bit more advanced. i.e. asking users to think through changing SLO parameters, more critical thinking type questions that SREs encounter.

Reply:
@nicoleczhu yes, the goal of the sandbox is to help a user to learn about SLOs and Cloud Ops services. Unfortunately, the existing UI does not support production grade SLOs and using simple SLOs does not teach too much about available options. I would argue that
# User has to use `gcloud` and not UI to do certain things with SLOs
# SLO does not provide "observability" about application problems. By default they raise alert about negative user experience

Regarding your suggestion in 1, if we were able to run the recipes in a virtual environment where everything is working except the one thing that we break, it make sense to take user directly to SLOs. Unfortunately, it is possible, especially if the load generator is working, there will be multiple issues with SLOs of different services. Without the first sentence, user my do a work unrelated to the recipe.

Can you please elaborate (2) and (3)? Where do you propose to put the explanation you mentioned? I appreciate if you can provide tangible example for more advanced quiz questions? I have them copy/pasted from two other recipes.

Reply:
Maybe something like this?

```
Product ratings are managed by the "rating service", hosted on Google AppEngine.
The service provides APIs that allow other services to get and update products' ratings.
The rating data is kept up-to-date by periodically calling an API endpoint that collects
all recently sent new rating scores for each product and calculates the new rating
based on the old value and the new scores. Try to check if the rating service operates normally.
```

I still have concerns the recipe may not be discoverable enough, but we had the same thoughts about some of the existing recipes too. I think after updating the hint, we can merge this as-is and review all the recipes later (before we start to publicize the recipes feature)